### PR TITLE
PlayerLaunchProjectileEvent

### PR DESCRIPTION
**Pull request by BillyGalbreath**
I'm pretty sure I got all player launched projectiles (except arrows). Let me know if I missed any.

This fixes a use-case specific issue that was discovered [here](https://www.spigotmc.org/threads/projectilehitevent-doesnt-work-in-survival-mode.327097/page-2#post-3059433). I have a use-case example a few posts down, [here](https://www.spigotmc.org/threads/projectilehitevent-doesnt-work-in-survival-mode.327097/page-2#post-3060204).

Not sure what to do about arrows since we have an EntityShootBowEvent already.

### CODE REVIEW

File: Spigot-API-Patches/0126-PlayerShootProjectileEvent.patch [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
++import org.bukkit.entity.Projectile;
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.entity.EntityEvent;
++import org.bukkit.inventory.ItemStack;
++
++/**
++ * Called when a player shoots a projectile
++ */
++public class PlayerShootProjectileEvent extends EntityEvent implements Cancellable {
```
Comment (aikar) [inline Spigot-API-Patches/0126-PlayerShootProjectileEvent.patch]: PlayerEvent

File: Spigot-API-Patches/0126-PlayerShootProjectileEvent.patch [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
++    private boolean cancelled;
++
++    public PlayerShootProjectileEvent(Player shooter, ItemStack itemStack, Projectile projectile) {
++        super(shooter);
++        this.itemStack = itemStack;
++        this.projectile = projectile;
++    }
++
++    @Override
++    public Player getEntity() {
```
Comment (aikar) [inline Spigot-API-Patches/0126-PlayerShootProjectileEvent.patch]: This isn't needed when extending PlayerEvent

File: Spigot-API-Patches/0126-PlayerShootProjectileEvent.patch [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
++        return (Player) entity;
++    }
++
++    /**
++     * Get the shooter of the projectile
++     *
++     * @return The shooter
++     */
++    public Player getShooter() {
```
Comment (aikar) [inline Spigot-API-Patches/0126-PlayerShootProjectileEvent.patch]: again becomes less useful with getPlayer(), PlayerShootItemEvent is clear enough getPlayer is the shooter, we don't need this method.

File: Spigot-API-Patches/0126-PlayerShootProjectileEvent.patch [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
++        return getEntity();
++    }
++
++    /**
++     * Gets the projectile which will be launched by this event
++     *
++     * @return the launched projectile
++     */
++    public Entity getProjectile() {
```
Comment (aikar) [inline Spigot-API-Patches/0126-PlayerShootProjectileEvent.patch]: return Projectile, not Entity

File: Spigot-API-Patches/0126-PlayerShootProjectileEvent.patch [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
++    public ItemStack getItemStack() {
++        return itemStack;
++    }
++
++    /**
++     * Get whether to consume the ItemStack or not
++     *
++     * @return True to consume
++     */
++    public boolean getConsumeStack() {
```
Comment (aikar) [inline Spigot-API-Patches/0126-PlayerShootProjectileEvent.patch]: I don't like this term. Why would it ever consume the entire stack?

Needs to be more clearly named of shouldConsume/setShouldConsume

File: Spigot-Server-Patches/0329-PlayerShootProjectileEvent.patch [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
++                } else ((EntityPlayer) entityhuman).getBukkitEntity().updateInventory();
++
++                world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.dL, SoundCategory.NEUTRAL, 0.5F, 0.4F / (ItemLingeringPotion.j.nextFloat() * 0.4F + 0.8F));
++            } else if (entityhuman instanceof EntityPlayer) {
++                ((EntityPlayer) entityhuman).getBukkitEntity().updateInventory();
++            }
++            // Paper end
+         }
+ 
+         entityhuman.b(StatisticList.b((Item) this));
```
Comment (aikar) [inline Spigot-Server-Patches/0329-PlayerShootProjectileEvent.patch]: I'm not sure we should fire statistics for cancelled events? same for the rest of the changed items.

**Review by aikar — CHANGES_REQUESTED** [review]

**Comment by aikar** [comment]
Also I think we should go with PlayerLaunchProjectileEvent

as you said, we have ShootBowEvent, you're not shooting these projectiles from anything.

Also, the API for launching them is clearly called .launchProjectile(), so lets go with that name.

**Comment by BillyGalbreath** [comment]
All fixed up

**Pull request closed without merging**